Resolved intermittent crash while seeking, added support for manifest files with CR-LF line breaks

neilrackett opened the pull request:
This pull request includes the following changes:
- `FragmentLoader.as` now checks for missing `_fragPrevious` data while analysing PTS to prevent crashes caused by a `TypeError: Error #1009: Cannot access a property or method of a null object reference` error during seek operations, which we have been seeing intermittently with some VOD streams
- A `numAudioTracks` getter has been added to `HLS.as` as a convenient way of checking for the availability of alternative audio (for example when displaying a menu, etc)
- Changes have been made to `Manifest.as`, and a new `StringUtil.as` class added, to support manifest files that have been encoded with line breaks or other whitespace characters not in the expected Linux format, for example those using CR-LF line breaks instead of just LF

neilrackett committed:
Added numAudioTracks getter

neilrackett committed:
Updated PTS analysis to resolve intermittent crash while seeking; Updated SWF/SWC;

neilrackett committed:
Resolved issues detecting media when loading manifest files that contain
lines with unexpected whitespace prefix or suffix

neilrackett committed:
Resolved issues loading non-Linux formatted manifest files

neilrackett committed:
Updated SWF/SWC

neilrackett committed:
Updated fix for non-Linux line breaks

mangui commented:
thanks @neilrackett , I am not comfortable with a PR with so many things inside.
best would be to split it into several PRs and also don't push swf/swc , i will take care of updating them.

mangui commented on .flexLibProperties:
```diff
@@ -16,6 +17,8 @@
     <classEntry path="org.mangui.osmf.plugins.traits.HLSTimeTrait"/>
     <classEntry path="org.mangui.osmf.plugins.utils.ErrorManager"/>
+    <classEntry path="org.mangui.hls.loader.SubtitlesFragmentLoader"/>
+    <classEntry path="org.mangui.hls.utils.hls_internal"/>
```
not existing files ?

neilrackett replied:
Removed (leftover from another branch)

mangui commented on src/org/mangui/hls/HLS.as:
```diff
@@ -259,7 +259,13 @@ package org.mangui.hls {
+        
+        /** How many audio tracks does the current media have? */
+        public function get numAudioTracks():uint
+        {
```
we can already get this info through hls.audioTracks, not sure it is that necessary to add another method here ?

neilrackett replied:
This information is only available via `hls.audioTracks` if alt audio tracks exist. This accessor is intended to remove the need to check that `hls.audioTracks` exists before checking the number of audio tracks available.

mangui replied:
ok then it is better to modify https://github.com/mangui/flashls/blob/dev/src/org/mangui/hls/loader/LevelLoader.as#L181 and initialize altAudioTracks to `new Vector.<AltAudioTrack>();` and remove this new method

neilrackett replied:
Done

mangui commented on src/org/mangui/hls/playlist/Manifest.as:
```diff
@@ -127,7 +128,7 @@ package org.mangui.hls.playlist {
         private function _loadCompleteHandler(event : Event) : void {
             _metrics.loading_end_time = getTimer();
-            onLoadedData(String(_urlloader.data));
+            onLoadedData(StringUtil.toLF(String(_urlloader.data)));
```
do you have a sample playlist which is not parsed correctly with current Manifest parser ?
I am just curious to see which kind of parsing issue you were facing.

neilrackett replied:
We faced a number of issues. but the primary problem was directives, such as `#EXT-X-MEDIA:`, being ignored by the parser because ActionScript reported them as starting at indexes of between 2 and 5, rather than 0.

Converting line breaks to LF and trimming whitespace resolved the issues and enabled the parser to parse as normal.

I'm not able to share URLs on a public forum, but here's a problematic playlist: [sample.m3u8](https://github.com/mangui/flashls/files/165748/sample.m3u8.txt)

mangui commented on src/org/mangui/hls/utils/StringUtil.as:
```diff
@@ -0,0 +1,107 @@
+		 * @returns			String without HTML tags
+		 * @example			StringUtil.removeHtmlTags("<p>Bonjour!</p>"); // Returns "Bonjour!"
+		 */
+		public static function removeHtmlTags(str:String):String
```
never used ?

neilrackett committed:
Removed references to classes that don't exist in this branch

neilrackett committed:
Removed StringUtil.as methods that are not used in this branch

neilrackett commented:
Hi @mangui, the PR started as a single item, but GitHub obviously includes all additional commits in the request, so I had to expand it -- I'll see what I can do to make PRs in more bite size chunks in the future (and won't commit SWF/SWC files).

I've removed class references and `StringUtil` methods not used in this branch, but let me know if there are any other tweaks to make before you can accept the PR.

neilrackett committed:
Merge remote-tracking branch 'upstream/dev' into dev

neilrackett committed:
Ensure all line breaks are LF before splitting into lines

neilrackett committed:
Merge remote-tracking branch 'upstream/dev' into dev

# Conflicts:
#	bin/debug/flashls.swc
#	bin/debug/flashlsChromeless.swf
#	bin/debug/flashlsFlowPlayer.swf
#	bin/debug/flashlsOSMF.swc
#	bin/debug/flashlsOSMF.swf
#	bin/release/flashls.swc
#	bin/release/flashlsChromeless.swf
#	bin/release/flashlsFlowPlayer.swf
#	bin/release/flashlsOSMF.swc
#	bin/release/flashlsOSMF.swf

neilrackett committed:
Ignored bin-temp

mangui commented on .actionScriptProperties:
```diff
@@ -1,6 +1,6 @@
 <?xml version="1.0" encoding="UTF-8" standalone="no"?>
 <actionScriptProperties analytics="false" mainApplicationPath="flashls.as" projectUUID="e938451b-f4d1-4796-ba95-b9054fce71fa" version="11">
-  <compiler additionalCompilerArguments="-locale=en_US&#13;&#10;-define+=CONFIG::LOGGING,false" autoRSLOrdering="true" copyDependentFiles="false" fteInMXComponents="false" generateAccessible="false" htmlExpressInstall="true" htmlGenerate="false" htmlHistoryManagement="false" htmlPlayerVersionCheck="true" includeNetmonSwc="false" outputFolderPath="bin" removeUnusedRSL="true" sourceFolderPath="src" ...
+  <compiler additionalCompilerArguments="-locale=en_US&#13;&#10;-define+=CONFIG::LOGGING,false" autoRSLOrdering="true" copyDependentFiles="false" fteInMXComponents="true" generateAccessible="false" htmlExpressInstall="true" htmlGenerate="false" htmlHistoryManagement="false" htmlPlayerVersionCheck="true" includeNetmonSwc="false" outputFolderPath="bin" removeUnusedRSL="true" sourceFolderPath="src"  ...
```
changes not related to the PR

neilrackett replied:
Reverted

mangui commented on .flexLibProperties:
```diff
@@ -1,6 +1,7 @@
 <?xml version="1.0" encoding="UTF-8" standalone="no"?>
-<flexLibProperties includeAllClasses="false" useMultiPlatformConfig="false" version="3">
+<flexLibProperties includeAllClasses="true" useMultiPlatformConfig="false" version="3">
```
changes not related to the PR

neilrackett replied:
Reverted

mangui commented on src/org/mangui/hls/playlist/Manifest.as:
```diff
@@ -387,7 +388,7 @@ package org.mangui.hls.playlist {
         public static function extractAltAudioTracks(data : String, base : String = '') : Vector.<AltAudioTrack> {
             var altAudioTracks : Vector.<AltAudioTrack> = new Vector.<AltAudioTrack>();
-            var lines : Array = data.split("\n");
+            var lines : Vector.<String> = StringUtil.toLines(data);
```
wouldn't this trim the Level / Alt Audio track name ?
it is really common to have space in Quality Level, track Name

neilrackett replied:
No: it trims whitespace characters from the ends of the lines, but has now affect on anything else, including labels

mangui replied:
ok got it

neilrackett committed:
Removed HLS.numAudioTracks and initialized
AudioTrackController.audioTracks using new Vector.<AltAudioTrack>() to
prevent crash when reading HLS.audioTracks.length before audio is ready

neilrackett committed:
Merge branch 'dev' of https://github.com/neilrackett/flashls into dev

neilrackett committed:
Reverted to settings used by mangui/flashls

mangui commented on src/org/mangui/hls/HLS.as:
```diff
@@ -259,7 +259,7 @@ package org.mangui.hls {
             _audioTrackController.audioTrack = val;
         }
-
+        
```
sorry  I am a bit pedantic, but there are some useless whitespace there :-/

neilrackett replied:
Not any more.

neilrackett committed:
Removed useless whitespace

neilrackett committed:
Removed useless whitespace

mangui commented:
thanks for your contribution ... and patience !
